### PR TITLE
Solid Earth Tide using per pixel Azimuth Time derived from ISCE2 orb.geo2rdr

### PR DESCRIPTION
**Pull request by cmarshak**
Solid Earth Tide computation using the `orb.geo2rdr` which outputs a tuple `azimuth_time` and `range` as noted [here](https://github.com/isce-framework/isce2/blob/c2d1a218020d81071574cd70033b7d673e2f5092/components/isceobj/Orbit/Orbit.py#L1000) shared by @mgovorcin.

Steps are:

1. Get azimuth time array with coordinate frame determined by `science/grids/imagingGeometry
2. Compute Solid Earth tide for the azimuth time rounded to nearest second above and below precise azimuth time and compute a weighted average.
3. Package into an xarray georeferenced dataset

The [notebook](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/SET-azimuth-time/tests/solid_earth_tides.ipynb) test is for interacting with the output and the pytest test file ensures georferencing is done correctly so can be ready by `gdal` or `rasterio`.

### CODE REVIEW

File: tests/test_set.py
Comment (cmarshak) [inline tests/test_set.py:58]: @dbekaert - I was able to show that in a vertical column of the azimuth time array that we comute has a standard deviation that never exceeds `1e-3` seconds.

However, if the heights in the `imagingGeometry` group have a max of 9,000 m and min of 1,500 m -- the speed of light takes 3.5e-5 seconds to travel between the highest and lowest level in the cube approximately (or 2 orders of magnitude smaller than the maximum standard deviation in this test).

Comment (dbekaert) [reply tests/test_set.py:58]: That sounds about right and negligible. We should just make a note of it in the code that this is the assumption we do. If we dont do it thats also ok. There overhead should be small.

**Comment by cmarshak** [comment]
Here are some sample products with overlapping spatial/temporal area: [GUNW 0](https://gunw-development-testing.s3.us-west-2.amazonaws.com/solid_earth_tide_may_2023/S1-GUNW-A-R-064-tops-20211120_20211108-014926-00119W_00032N-PP-c832-v2_0_6.nc) and [GUNW 1](https://gunw-development-testing.s3.us-west-2.amazonaws.com/solid_earth_tide_may_2023/S1-GUNW-A-R-064-tops-20211120_20211108-014951-00119W_00033N-PP-0643-v2_0_6.nc).

To update a product with the solid earth tide yourself, here is the sample code:

```
from pathlib import Path
import xarray as xr
from hyp3lib import get_orb
from isce2_topsapp.solid_earth_tides import update_gunw_with_solid_earth_tide

nc_path = Path('<GUNW_PATH>')
group = f'science/radarMetaData/inputSLC/reference'
with xr.open_dataset(nc_path, group=group) as ds:
    slc_id_ref = ds['L1InputGranules'].values[0]
group = f'science/radarMetaData/inputSLC/secondary'
with xr.open_dataset(nc_path, group=group) as ds:
    slc_id_sec = ds['L1InputGranules'].values[0]

orb_file_ref, _ = get_orb.downloadSentinelOrbitFile(slc_id_ref)
orb_file_sec, _ = get_orb.downloadSentinelOrbitFile(slc_id_sec)

update_gunw_with_solid_earth_tide(nc_path, 'reference', [orb_file_ref])
update_gunw_with_solid_earth_tide(nc_path, 'secondary', [orb_file_sec])
```

For the API, I have a list of orbit files, but I don't think that is necessary. May want to have single path or string.

@sssangha - take a look at it when you get back and then let Marin give a final review after you make any necessary modifications.

**Comment by cmarshak** [comment]
There is an issue that was discovered by @sssangha with this PR.

Unsure how this is normally dealt with.

It appears that `orb.geo2rdr` requires more steps to converge to the azimuth time and rng distance when using the *full* orbit file. In  ISCE3 there are options to increase the number of iterations of the numerical method to approximate this and in Raider there is a clipping of the orbit around is reference acquisition time. Because ISCE2 does not have options to increase the number of iterations we will pursue the latter.

**Comment by cmarshak** [comment]
@sssangha - this is ready for review - the same code snippet above still works.

I also added a test that I will link to below that resolves the issue you caught regarding geo2rdr.

File: tests/test_set.py
Comment (cmarshak) [inline tests/test_set.py:79]: @sssangha - ensures that the azimuth timing array never deviates from the first SLC start time by more than 70 seconds. For the GUNW under inspection, there are two secondary images (which are used above) and that's why we use the 70. Let me know if this looks good.

**Comment by sssangha** [comment]
@cmarshak I fixed the shift of lat array (previously the factor was added in this [line](https://github.com/ACCESS-Cloud-Based-InSAR/DockerizedTopsApp/blob/SET-azimuth-time/isce2_topsapp/solid_earth_tides.py#L177)) vis-a-vis the mintpy convention [here](https://github.com/insarlab/MintPy/blob/main/src/mintpy/prep_aria.py#L183).

**Comment by cmarshak** [comment]
@sssangha - I am reverting this.

Note in the mintpy is using the *affine transform* where we are using positive resolutions. The `+` was correct. Mintpy is using the signed resolutions as noted earlier in its code: https://github.com/insarlab/MintPy/blob/main/src/mintpy/prep_aria.py#L181

**Comment by cmarshak** [comment]
@sssangha - please approve once you are ready. After that, David can also review with you.

Thanks!

File: tests/test_set.py
Comment (cmarshak) [inline tests/test_set.py:170]: @dbekaert - FYI - this is a test that Sim and I worked on that ensures that the pysolid grid (used in mintpy) and the time varying interpolation used here do not vary by more than 1 mm across the scene.

File: tests/test_set.py
Comment (cmarshak) [inline tests/test_set.py:179]: @dbekaert - (had to remove previous comment because I updated the test) - this includes a verification that pysolid using fixed time does not deviate by more than 1 mm than the existing plugins method. This is something Sim and I worked on. This test will verify this for two datasets within the test suite and for both the reference and secondary dates in each dataset.

File: isce2_topsapp/solid_earth_tides.py
Comment (mgovorcin) [inline isce2_topsapp/solid_earth_tides.py:92]: wonder how is the padding set in the isce topsAp, is it also 600 seconds?

Comment (cmarshak) [reply isce2_topsapp/solid_earth_tides.py:92]: No idea...

Sim was able to track this down: https://github.com/dbekaert/RAiDER/blob/dev/tools/RAiDER/losreader.py#L188-L200

My rough understanding is that if there is not a padding then there are two solutions occurring at antipodal points on the globe.

File: isce2_topsapp/solid_earth_tides.py
Comment (mgovorcin) [inline isce2_topsapp/solid_earth_tides.py:177]: it think calculation of SET should be done for the pixel centroid rather than corner, not sure how pysolid calcs this as it accepts AREA style indeed. Is there anywhere in the code where pysolid does half pixel shift in both directions

**Review by mgovorcin — APPROVED** [review]
Looks great!, added two small comments for the discussion.

File: isce2_topsapp/solid_earth_tides.py
Comment (cmarshak) [reply isce2_topsapp/solid_earth_tides.py:177]: Not totally clear on what you are asking here.  

The python wrappers for the fortran code indeed want the `AREA` formatted metadata (as you mention) which would align with mintpy's attribute `dict`[https://github.com/insarlab/PySolid/blob/main/src/pysolid/grid.py#L30-L33]. I am assuming the georeference/gdal metadata of the outputs would align with those of the inputs (maybe this is where your question is?). This pysolid notebook is pretty clear about the inputs (not about the outputs): https://github.com/insarlab/PySolid/blob/main/docs/plot_grid_SET.ipynb

Also the other GUNW layers specifically the geometry rasters follow the convention:

1. Gdal metadata is `AREA`
2. The coordinates are `PIXEL`
3. The raster is consistent with 1.

This is what I have seen.

**Pull request merged**
